lib/workers.js

```diff
@@ -31,153 +33,156 @@ let pythonCallers = []; // all pythonCallers, whether available for work or curr
 let availablePythonCallers = []; // FIFO queue for available pythonCallers waiting for a job/callback
 let waitingCallbacks = []; // FIFO queue for jobs/callbacks waiting for an available pythonCaller
 
-module.exports = {
-    init() {
-        debug('init()');
-        if (!config.useWorkers) {
-            return;
-        }
-        var numWorkers = config.workersCount;
-        if (numWorkers == null) {
-            numWorkers = Math.ceil(config.workersPerCpu * os.cpus().length);
-        }
-        for (let i = 0; i < numWorkers; i++) {
-            const pc = new codeCaller.PythonCaller();
-            pc.number = i;
-            pythonCallers.push(pc);
-            load.startJob('python_worker_idle', pc.uuid);
-            availablePythonCallers.push(pc);
-        }
-        module.exports._warmUpWorkers();
-    },
-
-    finish(callback) {
-        debug('finish()');
-        if (!config.useWorkers) {
-            callback(null);
-            return;
-        }
-        const testFinished = () => {
-            if (availablePythonCallers.length < pythonCallers.length) {
-                // keep waiting for all pythonCallers to be returned
-                debug('finish(): waiting for availablePythonCallers');
-                setTimeout(testFinished, 100);
-            } else {
-                debug('finish(): calling done() on all pythonCallers');
-                for (let i = 0; i < pythonCallers.length; i++) {
-                    pythonCallers[i].done();
-                }
-                pythonCallers = [];
-                availablePythonCallers = [];
-                return callback(null);
+// Will be set to the appropriate class in init()
+let CodeCaller = null;
+
+module.exports.init = function() {
+    debug('init()');
+    const { workersExecutionMode } = config;
+    if (workersExecutionMode === 'container') {
+        CodeCaller = DockerCaller;
+    } else if (workersExecutionMode === 'native') {
+        CodeCaller = PythonCaller;
+    }
+    debug('using random allocation strategy');
+    let numWorkers = config.workersCount;
+    if (numWorkers == null) {
+        numWorkers = Math.ceil(config.workersPerCpu * os.cpus().length);
+    }
+
+    for (let i = 0; i < numWorkers; i++) {
+        const pc = new CodeCaller();
```

Comment:
Can we also track the `python` load variable here? This used to be tracked inside python caller, but now I think we need it here? This should be equal to the total number of code callers (active or idle).

Reply:
Check out 94a7bbe - I think that should do it.